lib/sshutils/x11/display.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package x11
+
+import (
+	"fmt"
+	"math"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	// DefaultDisplayOffset is the default display offset when
+	// searching for an open XServer unix socket.
+	DefaultDisplayOffset = 10
+	// DefaultMaxDisplay is the default maximum display number
+	// supported when searching for an open XServer unix socket.
+	DefaultMaxDisplay = 1000
+	// MaxDisplay is the theoretical max display value which
+	// X Clients and serverwill be able to parse into a unix socket.
+	MaxDisplayNumber = math.MaxInt32
+	// DisplayEnv is an environment variable used to determine what
+	// local display should be connected to during X11 forwarding.
+	DisplayEnv = "DISPLAY"
+
+	// x11BasePort is the base port used for XServer tcp addresses.
+	// e.g. DISPLAY=localhost:10 -> net.Dial("tcp", "localhost:6010")
+	// Used by some XServer clients, such as openSSH and MobaXTerm.
+	x11BasePort = 6000
+	// x11SocketDir is the name of the directory where X11 unix sockets are kept.
+	x11SocketDir = ".X11-unix"
+)
+
+// Display is an XServer display.
+type Display struct {
+	// HostName is the the display's hostname. For tcp display sockets, this will be
+	// an ip address. For unix display sockets, this will be empty or "unix".
+	HostName string `json:"hostname"`
+	// DisplayNumber is a number representing an x display.
+	DisplayNumber int `json:"display_number"`
+	// ScreenNumber is a specific screen number of an x display.
+	ScreenNumber int `json:"screen_number"`
+}
+
+// String returns the string representation of the display.
+func (d *Display) String() string {
+	return fmt.Sprintf("%s:%d.%d", d.HostName, d.DisplayNumber, d.ScreenNumber)
+}
+
+// Dial opens an XServer connection to the display
+func (d *Display) Dial() (XServerConn, error) {
+	tcpSock, tcpErr := d.tcpSocket()
+	if tcpErr == nil {
+		return net.DialTCP("tcp", nil, tcpSock)
```

Comment:
What happens if `d.hostName` is `"unix"`?

It looks like it would be possible for it to try `net.DialTCP`, fail, and never try `DialUnix`. Am I getting that right?

Reply:
I think `d.tcpSocket` would fail to resolve `unix:port`, but just in case of overlap I'll update Dial and Listen to try both and aggregate errors.